library/std/src/sys_common/fs.rs

```diff
@@ -32,11 +32,19 @@ pub fn remove_dir_all(path: &Path) -> io::Result<()> {
 
 fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
     for child in fs::read_dir(path)? {
-        let child = child?;
-        if child.file_type()?.is_dir() {
-            remove_dir_all_recursive(&child.path())?;
-        } else {
-            fs::remove_file(&child.path())?;
+        let result: io::Result<()> = try {
+            let child = child?;
+            if child.file_type()?.is_dir() {
+                remove_dir_all_recursive(&child.path())?;
+            } else {
+                fs::remove_file(&child.path())?;
+            }
+        };
+        // ignore internal NotFound errors to prevent race conditions
+        if let Err(err) = &result
+            && err.kind() != io::ErrorKind::NotFound
+        {
+            return result;
         }
     }
     fs::remove_dir(path)
```

Comment:
We should also ignore `NotFound` errors when deleting the directory after recursing through. Basically, the idea is that we only report `NotFound` if the specified directory didn't exist at the start of the call to `remove_dir_all`, and otherwise report a successful deletion even if someone else raced us to it.

Reply:
hmm, this initially seems like a good idea, but there's actually a lot of subtle issues i have:
1. the point of this PR is to trivialize implementing an idiomatic version of remove_dir_all  
2. ignoring this error may cause some race conditions to be more hidden, before two racing `remove_dir_all` would always cause one of them to give NotFound, but this tweak would cause such an error to occur only sometimes
3. this adds complexity to code that is duplicated 5 different times for various platforms, sometimes with the code subjected to complex conditional compilation, such as on macos, where the usual comptime check is actually deferred to runtime.

Reply:
I'm revisiting this after thinking about it a bit, and I still think we should ignore `NotFound` errors on the final removal.

My main argument is that we shouldn't return a `NotFound` error if we actually performed any deletions within the directory, since that means that the directory clearly was found. I would expect no deletions to have been done if a `NotFound` error is returned.

Reply:
yeah, it sounds like that's the consensus of t-libs on the ACP, and i hadn't considered the ambiguity of the error that would create.

i'll get working on it this morning.